CorsixTH/Src/th_gfx_sdl.cpp

```diff
@@ -389,6 +389,12 @@ bool render_target::set_scale_factor(double fScale, scaled_items eWhatToScale) {
     return false;
   } else if (eWhatToScale == scaled_items::all && direct_zoom) {
     global_scale_factor = fScale;
+    if ((SDL_GetWindowFlags(window) & SDL_WINDOW_FULLSCREEN_DESKTOP) ==
```

Comment:
This function is actually called even more than once a frame, at least 2, maybe 3 times. It's called with a value of 1 before drawing UI elements, and then with a value of the current zoom level before drawing the map. That's why flush_zoom_buffer is called at the beginning of this function and the end of the frame.

Reply:
> Do you have an idea why this works? I think I found the same thing years ago (why the fScale == 1 condition is below the zoom instead of above), but I don't recall why.

I haven't narrowed down exactly why the quality is so different. My leading guess is that drawing to an intermediate texure gives a pixel perfect rendering which is later upsampled as a whole whereas drawing directly is immediately upsampling each draw call smearing textures across pixels.

Reply:
> This function is actually called even more than once a frame, at least 2, maybe 3 times. It's called with a value of 1 before drawing UI elements, and then with a value of the current zoom level before drawing the map. That's why flush_zoom_buffer is called at the beginning of this function and the end of the frame.

It should be easy to remove the intermediate flush calls and use a single texture with direct_zoom given that the texture is always screen size.